biz.aQute.bndlib/src/aQute/bnd/osgi/MessageReporter.java

```diff
@@ -0,0 +1,279 @@
+			String message = this.message;
+
+			if (matcher != null && !matcher.isNegated()) {
+
+				Attrs attrs = fixupInstrs.get(matcher);
+				String restrict = attrs.get(Constants.FIXUPMESSAGES_RESTRICT_DIRECTIVE);
+				String replace = attrs.get(Constants.FIXUPMESSAGES_REPLACE_DIRECTIVE);
+				String is = attrs.get(Constants.FIXUPMESSAGES_IS_DIRECTIVE);
+
+				if (restrict == null || restrict.equals(type)) {
+
+					if (is != null && !is.equals(type)) {
+						error = !error;
+					}
+
+					if (replace != null) {
+						try (Processor replacer = new Processor(processor())) {
+							replacer.setProperty("@", message);
+							processor().getLogger()
+								.debug("replacing {} with {}", message, replace);
+							message = replacer.getReplacer()
+								.process(replace);
+						} catch (Exception e) {
+							throw Exceptions.duck(e);
+						}
+					}
+
+					if (attrs.isEmpty() || Constants.FIXUPMESSAGES_IS_IGNORE.equals(is)) {
+						message = null;
+					}
+				}
+			}
+			if (message != null) {
+				if (error)
+					errors.add(message);
+				else
+					warnings.add(message);
+			}
+		}
+
+	}
+
+	class Cache {
+		final List<String>	errors		= new ArrayList<>();
+		final List<String>	warnings	= new ArrayList<>();
+
+		Cache() {
+			boolean failOk = processor().isFailOk();
+			Instructions fixupInstrs = new Instructions();
+			Parameters fixup = processor().getMergedParameters(Constants.FIXUPMESSAGES);
+			fixup.forEach((k, v) -> fixupInstrs.put(Instruction.legacy(k), v));
+
+			messages.values()
+				.stream()
+				.sorted()
+				.forEach(m -> {
+					m.fixup(fixupInstrs, failOk ? warnings : errors, warnings);
+				});
+		}
+	}
+
+	MessageReporter(Processor processor) {
+		this.processor = processor;
+	}
+
+	List<String> getWarnings() {
+		return fixup().warnings;
+	}
+
+	List<String> getErrors() {
+		return fixup().errors;
+	}
+
+	Cache fixup() {
+		return new Cache();
+	}
+
+	Location getLocation(String msg) {
+		return messages.get(msg);
+	}
+
+	public SetLocation error(String format, Object... args) {
+		try {
+			for (int i = 0; i < args.length; i++) {
+				if (args[i] instanceof Throwable t) {
+					args[i] = Exceptions.causes(t);
+				}
+			}
+			String s = Processor.formatArrays(format, args);
+			Message m = new Message(counter.getAndIncrement(), s, true);
+			putMessage(s, m);
+			return m;
+		} finally {
+			processor().signal();
+		}
+	}
+
+	void putMessage(String s, Message m) {
+		ConcurrentHashMap<String, Message> current;
+		do {
+			current = messages;
+			current.putIfAbsent(s, m);
+		} while (current != messages);
+	}
+
+	SetLocation warning(String format, Object... args) {
+		String s = Processor.formatArrays(format, args);
+		Message m = new Message(counter.getAndIncrement(), s, false);
+		putMessage(s, m);
+		return m;
+	}
+
+	void getInfo(Reporter from, String prefix) {
+		String actualPrefix = prefix == null ? prefix = processor().getBase() + " :" : prefix;
```

Comment:
Don't use `prefix =` in the true side of the ternary operator. This is a non-obvious side effect.